spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/support/DefaultPublisherFactory.java

```diff
@@ -56,9 +59,17 @@
 
 	private final CredentialsProvider credentialsProvider;
 
-	public DefaultPublisherFactory(String projectId, ExecutorProvider executorProvider,
-			ChannelProvider channelProvider, CredentialsProvider credentialsProvider) {
-		this.projectId = projectId;
+	public DefaultPublisherFactory(GcpProjectIdProvider projectIdProvider,
+			ExecutorProvider executorProvider,
+			ChannelProvider channelProvider,
+			CredentialsProvider credentialsProvider) {
+		Assert.notNull(projectIdProvider, "The project ID provider can't be null.");
+		Assert.notNull(executorProvider, "The executor provider can't be null.");
+		Assert.notNull(channelProvider, "The channel provider can't be null.");
+		Assert.notNull(credentialsProvider, "The credentials provider can't be null.");
+
+		this.projectId = projectIdProvider.getProjectId()
+				.orElseThrow(MissingProjectIdException::new);
```

Comment:
The proper way to do that would be checking if `projectIdProvider.getProjectId()` is `null`. However, I thought we agreed that an `Optional` not being `null` is something we should accept as part of the contract?
If we do it for `projectIdProvider`, we should probably also do it for all the other providers.